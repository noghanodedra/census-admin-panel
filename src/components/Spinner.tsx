import React, { useContext, memo } from 'react';
import Loader from 'react-loader-spinner';

import { LoadingContext } from 'contexts';

const Spinner = ({ ...props }) => {
  const { loadingCount } = useContext(LoadingContext);

  return (
    <>
      {loadingCount > 0 && (
        <Loader
          type="Puff"
          color="#00BFFF"
          height={100}
          width={100}
          timeout={3000}
        />
      )}
    </>
  );
};

export default memo(Spinner);
