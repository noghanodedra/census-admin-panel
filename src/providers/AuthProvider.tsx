import React, {
  FunctionComponent, useState, useEffect,
} from 'react';
import { AuthContext } from 'contexts';

const AuthProvider: FunctionComponent = ({ children }) => {
  const [authData, setAuthData] = useState(null);

  // code for pre-loading the user's information if we have their token in

  // localStorage goes here

  // ðŸš¨ this is the important bit.

  // Normally your provider components render the context provider with a value.

  // But we post-pone rendering any of the children until after we've determined

  // whether or not we have a user token and if we do, then we render a spinner

  // while we go retrieve that user's information.

  // if (weAreStillWaitingToGetTheUserData) {
  //   return <FullPageSpinner />;
  // }

  useEffect(() => {
    // const currentAuthData = someManager.getAuthData();
    // if (currentAuthData) {
    //   setAuthData(currentAuthData);
    // }
  }, []);

  const onLogout = () => setAuthData(null);

  const onLogin = (newAuthData: object) => setAuthData(newAuthData);


  // note, I'm not bothering to optimize this `value` with React.useMemo here

  // because this is the top-most component rendered in our app and it will very

  // rarely re-render/cause a performance problem.

  return (
    <AuthContext.Provider
      value={{
        authData,
        onLogout,
        onLogin,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

const useAuth = () => React.useContext(AuthContext);

export { AuthProvider, useAuth };
