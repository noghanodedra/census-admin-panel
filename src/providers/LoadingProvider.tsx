import React, { useState, FunctionComponent } from 'react';
import { LoadingContext } from 'contexts';

const LoadingProvider: FunctionComponent = ({ children }) => {
  const showLoading = () => {
    toggleLoading((prevState) => ({
      ...prevState,
      loadingCount: prevState.loadingCount + 1,
    }));
  };

  const hideLoading = () => {
    toggleLoading((prevState) => ({
      ...prevState,
      loadingCount: prevState.loadingCount > 0
        ? prevState.loadingCount - 1 : 0,
    }));
  };

  const loadingState = {
    loadingCount: 0,
    showLoading,
    hideLoading,
  };

  const [loading, toggleLoading] = useState(loadingState);

  return (
    <LoadingContext.Provider value={loading}>
      {children}
    </LoadingContext.Provider>);
};

const useLoading = () => {
  const context = React.useContext(LoadingContext);
  if (context === undefined) {
    throw new Error('useLoading must be used within a AppProvider');
  }
  return context;
};

export { LoadingProvider, useLoading };
