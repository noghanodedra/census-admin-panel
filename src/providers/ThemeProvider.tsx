import React, { useState } from 'react';
import { MuiThemeProvider } from '@material-ui/core';
import { themeCreator } from 'themes/base';

export const ThemeContext = React.createContext((themeName: string): void => {});

const LIGHT_THEME = 'lightTheme';
const DARK_THEME = 'darkTheme';

const ThemeProvider: React.FC = (props) => {
  // Read current theme from localStorage or maybe from an api
  const curThemeName = localStorage.getItem('appTheme') || LIGHT_THEME;

  // State to hold the selected theme name
  const [themeName, _setThemeName] = useState(curThemeName);

  // Get the theme object by theme name
  console.log(themeName);
  const theme = themeCreator(themeName);

  const setThemeName = (thName: string): void => {
    localStorage.setItem('appTheme', thName);
    _setThemeName(themeName);
  };

  return (
    <ThemeContext.Provider value={setThemeName}>
      <MuiThemeProvider theme={theme}>{props.children}</MuiThemeProvider>
    </ThemeContext.Provider>
  );
};

export { ThemeProvider, LIGHT_THEME, DARK_THEME };
