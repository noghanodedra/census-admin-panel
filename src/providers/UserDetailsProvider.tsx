import React, { useState, FunctionComponent } from 'react';

import { USER_DETAILS_DEFAULT_VALUE, IUserDetails } from 'contexts/UserDetailsContext';
import { UserDetailsContext } from 'contexts';

const UserDetailsProvider: FunctionComponent = ({ children }) => {
  const userState = {
    ...USER_DETAILS_DEFAULT_VALUE,
  };

  const setUserDetails = (userDetails: IUserDetails) => {
    updateUserDetails((prevState) => ({ ...prevState, ...userDetails }));
  };

  const [userDetails, updateUserDetails] = useState(userState);

  return (
    <UserDetailsContext.Provider value={userDetails}>
      {children}
    </UserDetailsContext.Provider>);
};

const useUserDetails = () => {
  const context = React.useContext(UserDetailsContext);
  if (context === undefined) {
    throw new Error('useLoading must be used within a AppProvider');
  }
  return context;
};

export { UserDetailsProvider, useUserDetails };
