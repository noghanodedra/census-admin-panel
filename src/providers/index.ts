export { LoadingProvider } from './LoadingProvider';
export { ThemeProvider } from './ThemeProvider';
