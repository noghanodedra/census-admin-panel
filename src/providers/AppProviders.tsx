import React, { FunctionComponent } from 'react';
import { ApolloProvider } from '@apollo/react-hooks';

import makeApolloClient from 'utils/apollo';
import { LoadingProvider } from './LoadingProvider';
import { AuthProvider } from './AuthProvider';
import { ThemeProvider } from './ThemeProvider';
import { UserDetailsProvider } from './UserDetailsProvider';

const AppProviders: FunctionComponent = ({ children }) => {
  const client = makeApolloClient();

  return (
    <AuthProvider>
      <ApolloProvider client={client}>
        <ThemeProvider>
          <LoadingProvider>
            <UserDetailsProvider>
              {children}
            </UserDetailsProvider>
          </LoadingProvider>
        </ThemeProvider>
      </ApolloProvider>
    </AuthProvider>
  );
};
export default AppProviders;
