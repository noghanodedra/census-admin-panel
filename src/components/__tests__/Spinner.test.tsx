import React from 'react';
import { mount } from 'enzyme';
import { LoadingProvider } from 'providers';
import { useLoading } from 'providers/LoadingProvider';
import Spinner from '../Spinner';

describe('Render Spinner component', () => {
  it('render correctly ', () => {
    const SpinnerComponent = mount(<Spinner />);
    expect(SpinnerComponent).toMatchSnapshot();
  });

  xit('render correctly with context', () => {
    const tree = (
      <LoadingProvider>
        <Spinner />
      </LoadingProvider>
    );
    const SpinnerComponent = mount(tree);
    const { showLoading } = useLoading();
    showLoading();
    expect(SpinnerComponent).toMatchSnapshot();
  });
});
