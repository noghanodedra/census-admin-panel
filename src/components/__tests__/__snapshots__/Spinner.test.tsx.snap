// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Spinner component render correctly  1`] = `ReactWrapper {}`;

exports[`Render Spinner component render correctly with context 1`] = `ReactWrapper {}`;

exports[`Render Spinner render correctly Spinner component 1`] = `ReactWrapper {}`;
